pyramid_swagger/load_schema.py

```diff
@@ -35,190 +46,167 @@ def extract_query_param_schema(schema):
-                "$ref": "GeoPoint"
-            },
-        }
-    Which we can then validate against a JSON object we construct from the
-    pyramid request.
-    """
-    matching_body_schemas = [
-        s
-        for s in schema['parameters']
-        if s['paramType'] == 'body'
-    ]
-    if matching_body_schemas:
-        schema = matching_body_schemas[0]
-        type_ref = extract_validatable_type(schema['type'], models_schema)
-        # Unpleasant, but we are forced to replace 'type' defns with proper
-        # jsonschema refs.
-        if '$ref' in type_ref:
-            del schema['type']
-            schema.update(type_ref)
-        return strip_swagger_markup(schema)
-    else:
-        return None
+def noop_validator(_validator, *args):
+    return
 
 
-def strip_swagger_markup(schema):
-    """Turn a swagger URL parameter schema into a raw jsonschema.
+def build_swagger_type_validator(models):
+    def swagger_type_validator(validator, ref, instance, schema):
+        func = _validators.ref if ref in models else _validators.type_draft4
+        return func(validator, ref, instance, schema)
 
-    Involves just removing various swagger-specific markup tags.
-    """
-    swagger_specific_keys = (
-        'paramType',
-        'name',
-    )
-    return dict(
-        (k, v)
-        for k, v in schema.items()
-        if k not in swagger_specific_keys
-    )
+    return swagger_type_validator
 
 
-def get_model_resolver(schema):
+def required_validator(validator, req, instance, schema):
+    """Swagger 1.2 expects `required` to be a bool in the Parameter object, but
+    a list of properties in a Model object.
     """
-    Gets the schema and a RefResolver. RefResolver's will resolve "$ref:
-    ObjectType" entries in the schema, which are used to describe more complex
-    objects.
+    if schema.get('paramType') and req is True:
+        if not instance:
+            return [ValidationError("%s is required" % schema['name'])]
+    else:
+        return _validators.required_draft4(validator, req, instance, schema)
 
-    :returns: The RefResolver for the schema's models.
-    :rtype: RefResolver
-    """
-    models = dict(
-        (k, strip_swagger_markup(v))
-        for k, v in schema.get('models', {}).items()
+
+def get_body_validator(models):
+    return validators.extend(
+        Draft4Validator,
+        {
+            'paramType': noop_validator,
+            'name': noop_validator,
+            'type': build_swagger_type_validator(models),
+            'required': required_validator,
+        }
     )
-    return RefResolver('', '', models)
 
 
-class SchemaMap(namedtuple(
-        'SchemaMap', [
-            'request_query_schema',
-            'request_path_schema',
-            'request_body_schema',
-            'response_body_schema'
-        ])):
+Swagger12ParamValidator = validators.extend(
+    Draft3Validator,
+    {
+        'paramType': noop_validator,
+        'name': noop_validator,
+    }
+)
+
```

Comment:
This "use custom jsonschema.Validator for validation" stuff isn't really essential.  So if you think it's less-intuitive than the `strip_swagger_markup()`, I have no problem going back to that.

I do think it's kind of cool that we can create swagger validators out of jsonschema pretty easily. I'm not sure what the difference was between Draft3 and Draft4, but it's possible we could change this to Draft4 with other modifications.